File: mycelix-music/apps/web/src/lib/websocket.ts
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
/**
 * WebSocket Client for Mycelix Rust Backend
 *
 * Provides real-time communication for:
 * - Session synchronization
 * - Playback state updates
 * - Peer presence (collaborative listening)
 * - Live audio streaming
 */

const WS_BASE = process.env.NEXT_PUBLIC_WS_URL || 'ws://localhost:3000/ws';

// === Types ===

export interface SessionState {
  current_track?: string;
  position: number;
  playback: string;
  volume: number;
}

// Incoming messages (from server)
export type WsMessage =
  | { type: 'connected'; connection_id: string }
  | { type: 'session_created'; session_id: string }
  | { type: 'session_joined'; session_id: string; state: SessionState }
  | { type: 'playback_state_changed'; session_id: string; state: string }
  | { type: 'position_changed'; session_id: string; position: number }
  | { type: 'volume_changed'; session_id: string; volume: number }
  | { type: 'track_changed'; session_id: string; track_id: string }
  | { type: 'peer_joined'; session_id: string; peer_id: string }
  | { type: 'peer_left'; session_id: string; peer_id: string }
  | { type: 'subscribed'; session_id: string; events: string[] }
  | { type: 'audio_data'; session_id: string; timestamp: number; data: string }
  | { type: 'error'; message: string }
  | { type: 'pong'; data: number[] };

// Outgoing commands (to server)
export type WsCommand =
  | { type: 'create_session' }
  | { type: 'join_session'; session_id: string }
  | { type: 'play'; session_id: string }
  | { type: 'pause'; session_id: string }
  | { type: 'seek'; session_id: string; position: number }
  | { type: 'set_volume'; session_id: string; volume: number }
  | { type: 'subscribe_to_session'; session_id: string }
  | { type: 'ping' };

// Event handlers
export interface WsEventHandlers {
  onConnected?: (connectionId: string) => void;
  onSessionCreated?: (sessionId: string) => void;
  onSessionJoined?: (sessionId: string, state: SessionState) => void;
  onPlaybackStateChanged?: (sessionId: string, state: string) => void;
  onPositionChanged?: (sessionId: string, position: number) => void;
  onVolumeChanged?: (sessionId: string, volume: number) => void;
  onTrackChanged?: (sessionId: string, trackId: string) => void;
  onPeerJoined?: (sessionId: string, peerId: string) => void;
  onPeerLeft?: (sessionId: string, peerId: string) => void;
  onSubscribed?: (sessionId: string, events: string[]) => void;
  onAudioData?: (sessionId: string, timestamp: number, data: ArrayBuffer) => void;
  onError?: (message: string) => void;
  onDisconnected?: () => void;
  onReconnecting?: (attempt: number) => void;
}

// Connection options
export interface WsOptions {
  url?: string;
  reconnect?: boolean;
  reconnectAttempts?: number;
  reconnectDelay?: number;
  pingInterval?: number;
}

// === WebSocket Client Class ===

export class MycelixWebSocket {
  private ws: WebSocket | null = null;
  private url: string;
  private handlers: WsEventHandlers = {};
  private reconnect: boolean;
  private reconnectAttempts: number;
  private reconnectDelay: number;
  private reconnectCount = 0;
  private pingInterval: number;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private connectionId: string | null = null;

  constructor(options: WsOptions = {}) {
    this.url = options.url || WS_BASE;
    this.reconnect = options.reconnect ?? true;
    this.reconnectAttempts = options.reconnectAttempts ?? 5;
    this.reconnectDelay = options.reconnectDelay ?? 1000;
    this.pingInterval = options.pingInterval ?? 30000;
  }

  /**
   * Set event handlers
   */
  on(handlers: WsEventHandlers): this {
    this.handlers = { ...this.handlers, ...handlers };
    return this;
  }

  /**
   * Connect to WebSocket server
   */
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.url);

        this.ws.onopen = () => {
          this.reconnectCount = 0;
          this.startPing();
          resolve();
        };

        this.ws.onmessage = (event) => {
          this.handleMessage(event.data);
        };

        this.ws.onerror = (error) => {
          console.error('WebSocket error:', error);
          reject(error);
        };

        this.ws.onclose = () => {
          this.stopPing();
          this.connectionId = null;
          this.handlers.onDisconnected?.();

          if (this.reconnect && this.reconnectCount < this.reconnectAttempts) {
            this.reconnectCount++;
            this.handlers.onReconnecting?.(this.reconnectCount);
            setTimeout(() => this.connect(), this.reconnectDelay * this.reconnectCount);
          }
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Disconnect from server
   */
  disconnect(): void {
    this.reconnect = false;
    this.stopPing();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Check if connected
   */
  get isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  /**
   * Get connection ID
   */
  get id(): string | null {
    return this.connectionId;
  }

  // --- Commands ---

  /**
   * Send a command to the server
   */
  send(command: WsCommand): void {
    if (!this.isConnected) {
      console.error('WebSocket not connected');
      return;
    }

    this.ws!.send(JSON.stringify(command));
  }

  /**
   * Create a new session
   */
  createSession(): void {
    this.send({ type: 'create_session' });
  }

  /**
   * Join an existing session
   */
  joinSession(sessionId: string): void {
    this.send({ type: 'join_session', session_id: sessionId });
  }

  /**
   * Play in session
   */
  play(sessionId: string): void {
    this.send({ type: 'play', session_id: sessionId });
  }

  /**
   * Pause in session
   */
  pause(sessionId: string): void {
    this.send({ type: 'pause', session_id: sessionId });
  }

  /**
   * Seek in session
   */
  seek(sessionId: string, position: number): void {
    this.send({ type: 'seek', session_id: sessionId, position });
  }

  /**
   * Set volume in session
   */
  setVolume(sessionId: string, volume: number): void {
    this.send({ type: 'set_volume', session_id: sessionId, volume });
  }

  /**
   * Subscribe to session events
   */
  subscribeToSession(sessionId: string): void {
    this.send({ type: 'subscribe_to_session', session_id: sessionId });
  }

  /**
   * Send ping (heartbeat)
   */
  ping(): void {
    this.send({ type: 'ping' });
  }

  // --- Private Methods ---

  private handleMessage(data: string): void {
    try {
      const message: WsMessage = JSON.parse(data);

      switch (message.type) {
        case 'connected':
          this.connectionId = message.connection_id;
          this.handlers.onConnected?.(message.connection_id);
          break;

        case 'session_created':
          this.handlers.onSessionCreated?.(message.session_id);
          break;

        case 'session_joined':
          this.handlers.onSessionJoined?.(message.session_id, message.state);
          break;

        case 'playback_state_changed':
          this.handlers.onPlaybackStateChanged?.(message.session_id, message.state);
          break;

        case 'position_changed':
          this.handlers.onPositionChanged?.(message.session_id, message.position);
          break;

        case 'volume_changed':
          this.handlers.onVolumeChanged?.(message.session_id, message.volume);
          break;

        case 'track_changed':
          this.handlers.onTrackChanged?.(message.session_id, message.track_id);
          break;

        case 'peer_joined':
          this.handlers.onPeerJoined?.(message.session_id, message.peer_id);
          break;

        case 'peer_left':
          this.handlers.onPeerLeft?.(message.session_id, message.peer_id);
          break;

        case 'subscribed':
          this.handlers.onSubscribed?.(message.session_id, message.events);
          break;

        case 'audio_data':
          // Decode base64 audio data
          const binary = atob(message.data);
          const bytes = new Uint8Array(binary.length);
          for (let i = 0; i < binary.length; i++) {
            bytes[i] = binary.charCodeAt(i);
          }
          this.handlers.onAudioData?.(message.session_id, message.timestamp, bytes.buffer);
          break;

        case 'error':
          this.handlers.onError?.(message.message);
          break;

        case 'pong':
          // Heartbeat response, ignore
          break;
      }
    } catch (error) {
      console.error('Failed to parse WebSocket message:', error);
    }
  }

  private startPing(): void {
    this.stopPing();
    this.pingTimer = setInterval(() => {
      if (this.isConnected) {
        this.ping();
      }
    }, this.pingInterval);
  }

  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }
}

// === Singleton Instance ===

let wsInstance: MycelixWebSocket | null = null;

/**
 * Get or create the WebSocket instance
 */
export function getWebSocket(options?: WsOptions): MycelixWebSocket {
  if (!wsInstance) {
    wsInstance = new MycelixWebSocket(options);
  }
  return wsInstance;
}

/**
 * Connect the singleton WebSocket
 */
export async function connectWebSocket(options?: WsOptions): Promise<MycelixWebSocket> {
  const ws = getWebSocket(options);
  if (!ws.isConnected) {
    await ws.connect();
  }
  return ws;
}

/**
 * Disconnect the singleton WebSocket
 */
export function disconnectWebSocket(): void {
  if (wsInstance) {
    wsInstance.disconnect();
    wsInstance = null;
  }
}

export default MycelixWebSocket;

File: mycelix-workspace/mycelix-music/rust/crates/mycelix-api/src/websocket.rs
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
//! WebSocket handlers for real-time communication

use std::sync::Arc;

use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        State,
    },
    response::IntoResponse,
};
use futures::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;
use uuid::Uuid;

use crate::AppState;

/// WebSocket upgrade handler
pub async fn ws_handler(
    ws: WebSocketUpgrade,
    State(state): State<AppState>,
) -> impl IntoResponse {
    ws.on_upgrade(move |socket| handle_socket(socket, state))
}

/// Handle WebSocket connection
async fn handle_socket(socket: WebSocket, state: AppState) {
    let (mut sender, mut receiver) = socket.split();

    // Create channel for sending messages to client
    let (tx, mut rx) = mpsc::channel::<WsMessage>(32);

    // Connection ID
    let connection_id = Uuid::new_v4();
    tracing::info!("WebSocket connected: {}", connection_id);

    // Spawn task to forward messages to WebSocket
    let send_task = tokio::spawn(async move {
        while let Some(msg) = rx.recv().await {
            let json = serde_json::to_string(&msg).unwrap_or_default();
            if sender.send(Message::Text(json)).await.is_err() {
                break;
            }
        }
    });

    // Send welcome message
    let _ = tx
        .send(WsMessage::Connected {
            connection_id: connection_id.to_string(),
        })
        .await;

    // Handle incoming messages
    while let Some(msg) = receiver.next().await {
        match msg {
            Ok(Message::Text(text)) => {
                if let Ok(cmd) = serde_json::from_str::<WsCommand>(&text) {
                    handle_command(cmd, &state, &tx).await;
                }
            }
            Ok(Message::Binary(data)) => {
                // Handle binary audio data
                tracing::debug!("Received {} bytes of binary data", data.len());
            }
            Ok(Message::Ping(data)) => {
                let _ = tx.send(WsMessage::Pong { data }).await;
            }
            Ok(Message::Close(_)) => {
                tracing::info!("WebSocket closed: {}", connection_id);
                break;
            }
            Err(e) => {
                tracing::error!("WebSocket error: {}", e);
                break;
            }
            _ => {}
        }
    }

    send_task.abort();
    tracing::info!("WebSocket disconnected: {}", connection_id);
}

/// Handle incoming WebSocket commands
async fn handle_command(cmd: WsCommand, state: &AppState, tx: &mpsc::Sender<WsMessage>) {
    match cmd {
        WsCommand::CreateSession => {
            match state.engine.create_session().await {
                Ok(session_id) => {
                    let _ = tx
                        .send(WsMessage::SessionCreated {
                            session_id: session_id.to_string(),
                        })
                        .await;
                }
                Err(e) => {
                    let _ = tx
                        .send(WsMessage::Error {
                            message: e.to_string(),
                        })
                        .await;
                }
            }
        }

        WsCommand::JoinSession { session_id } => {
            match Uuid::parse_str(&session_id) {
                Ok(uuid) => match state.engine.get_session(uuid) {
                    Ok(session) => {
                        let session_state = session.state();
                        let _ = tx
                            .send(WsMessage::SessionJoined {
                                session_id,
                                state: SessionStateMsg {
                                    current_track: session_state.current_track.map(|id| id.to_string()),
                                    position: session_state.position,
                                    playback: format!("{:?}", session_state.playback),
                                    volume: session_state.volume,
                                },
                            })
                            .await;
                    }
                    Err(e) => {
                        let _ = tx
                            .send(WsMessage::Error {
                                message: e.to_string(),
                            })
                            .await;
                    }
                },
                Err(_) => {
                    let _ = tx
                        .send(WsMessage::Error {
                            message: "Invalid session ID".to_string(),
                        })
                        .await;
                }
            }
        }

        WsCommand::Play { session_id } => {
            if let Ok(uuid) = Uuid::parse_str(&session_id) {
                if let Ok(session) = state.engine.get_session(uuid) {
                    session.play();
                    let _ = tx
                        .send(WsMessage::PlaybackStateChanged {
                            session_id,
                            state: "Playing".to_string(),
                        })
                        .await;
                }
            }
        }

        WsCommand::Pause { session_id } => {
            if let Ok(uuid) = Uuid::parse_str(&session_id) {
                if let Ok(session) = state.engine.get_session(uuid) {
                    session.pause();
                    let _ = tx
                        .send(WsMessage::PlaybackStateChanged {
                            session_id,
                            state: "Paused".to_string(),
                        })
                        .await;
                }
            }
        }

        WsCommand::Seek { session_id, position } => {
            if let Ok(uuid) = Uuid::parse_str(&session_id) {
                if let Ok(session) = state.engine.get_session(uuid) {
                    session.seek(position);
                    let _ = tx
                        .send(WsMessage::PositionChanged {
                            session_id,
                            position,
                        })
                        .await;
                }
            }
        }

        WsCommand::SetVolume { session_id, volume } => {
            if let Ok(uuid) = Uuid::parse_str(&session_id) {
                if let Ok(session) = state.engine.get_session(uuid) {
                    session.set_volume(volume);
                    let _ = tx
                        .send(WsMessage::VolumeChanged {
                            session_id,
                            volume,
                        })
                        .await;
                }
            }
        }

        WsCommand::SubscribeToSession { session_id } => {
            // In production, would set up event streaming for this session
            let _ = tx
                .send(WsMessage::Subscribed {
                    session_id,
                    events: vec![
                        "playback_state".to_string(),
                        "position".to_string(),
                        "track_changed".to_string(),
                        "peer_joined".to_string(),
                        "peer_left".to_string(),
                    ],
                })
                .await;
        }

        WsCommand::Ping => {
            let _ = tx.send(WsMessage::Pong { data: vec![] }).await;
        }
    }
}

/// Incoming WebSocket commands
#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WsCommand {
    CreateSession,
    JoinSession { session_id: String },
    Play { session_id: String },
    Pause { session_id: String },
    Seek { session_id: String, position: f32 },
    SetVolume { session_id: String, volume: f32 },
    SubscribeToSession { session_id: String },
    Ping,
}

/// Outgoing WebSocket messages
#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WsMessage {
    Connected {
        connection_id: String,
    },
    SessionCreated {
        session_id: String,
    },
    SessionJoined {
        session_id: String,
        state: SessionStateMsg,
    },
    PlaybackStateChanged {
        session_id: String,
        state: String,
    },
    PositionChanged {
        session_id: String,
        position: f32,
    },
    VolumeChanged {
        session_id: String,
        volume: f32,
    },
    TrackChanged {
        session_id: String,
        track_id: String,
    },
    PeerJoined {
        session_id: String,
        peer_id: String,
    },
    PeerLeft {
        session_id: String,
        peer_id: String,
    },
    Subscribed {
        session_id: String,
        events: Vec<String>,
    },
    AudioData {
        session_id: String,
        timestamp: f64,
        #[serde(with = "base64_serde")]
        data: Vec<u8>,
    },
    Error {
        message: String,
    },
    Pong {
        data: Vec<u8>,
    },
}

#[derive(Debug, Serialize)]
pub struct SessionStateMsg {
    pub current_track: Option<String>,
    pub position: f32,
    pub playback: String,
    pub volume: f32,
}

/// Base64 serialization for binary data
mod base64_serde {
    use base64::{engine::general_purpose::STANDARD, Engine};
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S>(bytes: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&STANDARD.encode(bytes))
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        STANDARD.decode(&s).map_err(serde::de::Error::custom)
    }
}
